File: src/sparsify.jl
# Adaptive spectral sparsifier + options + SparseMatrixCSC<->edge bridge.
#
# port: CMG-python src/pycmg/_sparsify.py (SparsifyOptions, spanner_bundle,
#       sparsify, edges_of, slack_of, sdd_from_edges).
#
# The sparsifier keeps a spanner (the "bundle") at weight 1 and keeps each
# off-bundle edge with a single probability `p` chosen to hit a target
# keep-fraction, reweighting survivors by `1/p` (an unbiased estimator of the
# off-bundle part). Because a spanner bounds off-spanner leverage, this uniform
# sampling concentrates.

"""
    SparsifyOptions

Tuning knobs for sparsify-on-stall (see `cmg_solve(A, b; sparsify_on_stall=true,
sparsify_opts=SparsifyOptions(...))`). Defaults are the production values
validated in the CMG-python package.

- `keep_frac`  : target keep-fraction for off-bundle edges (0.5 -> C_op ~ 2).
                 For the L-cycle this sets the injected level's repeat multiplier
                 = floor(1/keep_frac - 1): 0.5 -> repeat 1 (cheapest), 1/3 ->
                 repeat 2. Lower keep_frac sparsifies harder but raises the repeat.
- `bundles`    : number of peeled spanners/forests kept at full weight (default
                 3, i.e. a 3-forest Nagamochi–Ibaraki bundle for `:mst`). More
                 bundles -> lower effective stretch / higher cut connectivity;
                 each is an extra forest build (cheap for `:mst`: O(m) each).
- `t`          : greedy-spanner stretch; `nothing` -> `max(2, log2 n)` per level.
- `stall_ratio`: per-level NODE-coarsening threshold; a level is productive iff
                 aggregation drops the node count below `stall_ratio * n`
                 (`nc > stall_ratio * n` is a stall -> sparsify). This is the
                 validated CMG-python criterion; the original Julia port
                 mistakenly tested edges (`m_c <= stall_ratio*m`), which fired on
                 normal levels -- corrected to nodes here.
- `max_inject` : cap on injected sparsifier levels (guarantees termination).
- `nnz_budget` : cumulative operator-complexity budget, a multiple of input nnz
                 (the same 5x guard stock CMG uses). `Inf` disables it.
- `spanner`    : `:mst` (default; a `bundles`-forest maximum-spanning-forest
                 bundle — a Nagamochi–Ibaraki cut certificate, O(m) to build, and
                 on the high-conductance stall levels where sparsify fires cut ≈
                 spectral so it is a good spectral preconditioner there at a
                 fraction of Baswana–Sen's cost — see `max_spanning_forest`),
                 `:baswana_sen` (the randomized (2k-1)-spanner), or `:greedy`.
- `k`          : Baswana–Sen stretch parameter; `nothing` -> `ceil(log2 n)`.
- `bundles_growth` : grow the bundle count by one per successive injection.
- `base`       : direct-solve threshold (matches `build_hierarchy`'s 500).
- `seed`       : RNG seed (the spanner + sampling are randomized; seeding makes
                 the preconditioner reproducible).
"""
Base.@kwdef struct SparsifyOptions
    keep_frac::Float64 = 0.5
    bundles::Int = 3
    t::Union{Nothing,Float64} = nothing
    stall_ratio::Float64 = 0.9   # NODE-coarsening threshold: a level is
                                 # productive iff aggregation drops nc below
                                 # stall_ratio*n; nc > stall_ratio*n is a stall.
    max_inject::Int = 10
    nnz_budget::Float64 = 5.0
    spanner::Symbol = :mst
    k::Union{Nothing,Int} = nothing
    bundles_growth::Bool = false
    base::Int = 500
    seed::Int = 0
end

# ---------------------------------------------------------------------------
# spanner bundle + adaptive sparsify
# ---------------------------------------------------------------------------
"""
    spanner_bundle(n, edges, t, bundles; spanner, k, rng) -> (bundle, off)

Peel `bundles` spanners: extract a spanner, remove its edges, repeat. Returns
`(bundle_edges, off_bundle_edges)`. More bundles -> lower effective stretch ->
aggregation resumes harder at the same edge budget.
"""
function spanner_bundle(n::Integer, edges::AbstractVector, t::Real, bundles::Integer;
                        spanner::Symbol = :baswana_sen, k = nothing,
                        rng::AbstractRNG = Random.default_rng())
    _one(rem) = spanner === :greedy      ? greedy_spanner(n, rem, t) :
                spanner === :baswana_sen ? spanner_baswana_sen(n, rem; k = k, rng = rng) :
                spanner === :mst         ? max_spanning_forest(n, rem) :
                error("unknown spanner $(repr(spanner)); use :greedy, :baswana_sen, or :mst")
    ekey(e) = (min(e[1], e[2]), max(e[1], e[2]))
    kept = SpanEdge[]
    rem = collect(SpanEdge, edges)
    for _ in 1:bundles
        s = _one(rem)
        append!(kept, s)
        sset = Set(ekey(e) for e in s)
        rem = [e for e in rem if !(ekey(e) in sset)]
    end
    return kept, rem
end

"""
    sparsify(n, edges; t, bundles, keep_frac, spanner, k, rng) -> (kept, p)

`t=nothing` -> `t = max(2, log2 n)`. `p = clamp((keep_frac*m - S)/(m - S), 0, 1)`
with `S = |bundle|`, `m = |edges|`; the bundle is kept at weight 1 and each
off-bundle edge is kept w.p. `p` and reweighted `w/p`. A sparse level (`S ~ m`)
drives `p -> 0`; a dense/fill level (`S << m`) gives `p ~ keep_frac`.
"""
function sparsify(n::Integer, edges::AbstractVector; t = nothing, bundles::Integer = 1,
                  keep_frac::Real = 0.5, spanner::Symbol = :baswana_sen, k = nothing,
                  rng::AbstractRNG = Random.default_rng())
    tt = t === nothing ? (n > 1 ? max(2.0, log2(n)) : 2.0) : float(t)
    bundle, off = spanner_bundle(n, edges, tt, bundles; spanner = spanner, k = k, rng = rng)
    m, S = length(edges), length(bundle)
    (m <= S || isempty(off)) && return collect(SpanEdge, edges), 0.0
    p = min(1.0, max(0.0, (keep_frac * m - S) / (m - S)))
    kept = copy(bundle)
    if p > 0.0
        @inbounds for (u, v, w) in off
            if rand(rng) < p
                push!(kept, (u, v, w / p))
            end
        end
    end
    return kept, p
end

# ---------------------------------------------------------------------------
# SparseMatrixCSC <-> edge-list bridge (SDD ops: L(graph) + diag(slack), >= 0)
# ---------------------------------------------------------------------------
"""
    edges_of(A) -> Vector{Tuple{Int,Int,Float64}}

Off-diagonal edges `(u, v, w)` with `u < v`, `w = -A[u,v] > 0`, from a sparse
SDD/Laplacian operator. Explicit zeros are skipped (not edges).
"""
function edges_of(A::SparseMatrixCSC)
    edges = SpanEdge[]
    rows = rowvals(A); vals = nonzeros(A)
    @inbounds for j in 1:size(A, 2)
        for r in nzrange(A, j)
            i = rows[r]
            if i < j                                # strict upper triangle
                val = vals[r]
                val != 0.0 && push!(edges, (Int64(i), Int64(j), -float(val)))
            end
        end
    end
    return edges
end

"""
    slack_of(A) -> Vector{Float64}

Per-row SDD slack = `diag(A) - (sum of incident conductances)`, which for a
symmetric SDD/Laplacian equals the row sums of `A`: `>= 0` for SDD, `~0` for a
Laplacian. Preserving it keeps a sparsified level SPD-consistent with its
parent; at every internal CMG level (a Laplacian) it is ~0, so the sparsifier
there is a pure Laplacian with the same null space.
"""
slack_of(A::SparseMatrixCSC) = vec(sum(A, dims = 2))

"""
    sdd_from_edges(n, edges, slack) -> SparseMatrixCSC

Build `L(edges) + diag(slack)`. `slack` may be a scalar or a length-`n` vector
(e.g. `slack_of(A)`).
"""
function sdd_from_edges(n::Integer, edges::AbstractVector, slack)
    deg = zeros(Float64, n)
    ne = length(edges)
    I = Vector{Int64}(undef, 2 * ne)
    J = Vector{Int64}(undef, 2 * ne)
    V = Vector{Float64}(undef, 2 * ne)
    @inbounds for (idx, e) in enumerate(edges)
        u, v, w = Int64(e[1]), Int64(e[2]), float(e[3])
        I[2idx - 1] = u; J[2idx - 1] = v; V[2idx - 1] = -w
        I[2idx]     = v; J[2idx]     = u; V[2idx]     = -w
        deg[u] += w; deg[v] += w
    end
    slackv = slack isa Number ? fill(float(slack), n) : collect(Float64, slack)
    dI = collect(Int64, 1:n)
    A = sparse(vcat(I, dI), vcat(J, dI), vcat(V, deg .+ slackv), Int(n), Int(n))
    return A
end

File: experiments/sparsified/sparsify.py
#!/usr/bin/env python3
"""
Isolated spanner + adaptive spectral sparsifier for sparsified-CMG.

Self-contained numeric module (NO cmg-python imports) so it ports 1:1 to
Julia (spanner.jl / sparsify.jl). Edges are `(u, v, w)` tuples with w > 0 a
conductance and u < v -- identical to reference.py, so `greedy_spanner`,
`spanner_bundle`, and `sparsify` are line-for-line the reference algorithm.
The only additions over reference.py are:

  * `t` defaults to `log2(n)` (reference.py hardcodes t=9.0 for n~400-960,
    which IS log2(n) there, NOT ln(n) -- with ln(n) the greedy spanner keeps
    too many edges and the 0.98 reduction gate never lets the fork fire);
  * a scipy<->edge-list bridge (`edges_of`, `slack_of`, `sdd_from_edges`) that
    replaces reference.py's dense `lap_dense`/`edges_of`/`slack_of`, since the
    real CMG operators are sparse.

See experiments/sparsified/PYTHON-GUIDELINES.md and reference.py.
"""
import heapq
import math
import random

import numpy as np
import scipy.sparse as sp


# ---------------------------------------------------------------------------
# spanner (greedy, resistance metric: edge length = 1/conductance) + bundle
# ---------------------------------------------------------------------------
def greedy_spanner(n, edges, t):
    """Greedy (Althofer) t-spanner in the resistance metric.

    Keep edge (u, v) unless the current spanner already connects u, v with
    resistance <= t / w_uv. A t-spanner bounds every off-spanner edge's
    effective resistance by t * r_uv, i.e. leverage <= t, which is what makes
    uniform sampling concentrate. `edges`: list of (u, v, w) conductances.
    """
    adj = [[] for _ in range(n)]
    span = []

    def within(s, tgt, cap):
        # bounded-radius Dijkstra: is there an s->tgt spanner path of length <= cap?
        d = {s: 0.0}
        pq = [(0.0, s)]
        while pq:
            du, u = heapq.heappop(pq)
            if u == tgt:
                return du <= cap
            if du > cap:
                return False
            if du > d.get(u, 1e18):
                continue
            for (v, l) in adj[u]:
                nd = du + l
                if nd < d.get(v, 1e18):
                    d[v] = nd
                    heapq.heappush(pq, (nd, v))
        return False

    for (u, v, w) in sorted(edges, key=lambda e: 1.0 / e[2]):   # stiffest first
        L = 1.0 / w
        if not within(u, v, t * L):
            adj[u].append((v, L))
            adj[v].append((u, L))
            span.append((u, v, w))
    return span


def spanner_bundle(n, edges, t, bundles):
    """Peel `bundles` spanners: extract a spanner, remove its edges, repeat.
    Returns (bundle_edges, off_bundle_edges). More bundles -> lower effective
    stretch -> aggregation resumes harder at the same edge budget."""
    key = lambda e: (min(e[0], e[1]), max(e[0], e[1]))
    kept = []
    rem = list(edges)
    for _ in range(bundles):
        s = greedy_spanner(n, rem, t)
        kept += s
        sset = set(key(e) for e in s)
        rem = [e for e in rem if key(e) not in sset]
    return kept, rem


# ---------------------------------------------------------------------------
# adaptive sparsifier: keep the bundle at weight 1; sample off-bundle edges at
# a probability CHOSEN from the bundle size to hit a target keep-fraction.
# ---------------------------------------------------------------------------
def sparsify(n, edges, t=None, bundles=1, keep_frac=0.5, spanner="greedy",
             rng=None):
    """Return (kept_edges, p). `t=None` -> t = max(2.0, log2(n)).

    p = clamp((keep_frac*m - S)/(m - S), 0, 1); the bundle (S edges) is kept at
    weight 1, each off-bundle edge is kept w.p. p and reweighted by 1/p (an
    unbiased estimator of the off-bundle part). A sparse level (S ~ m) drives
    p -> 0 (keep just the bundle, no over-sparsifying); a dense/fill level
    (S << m) gives p ~ keep_frac (a gentle geometric drop).
    """
    if t is None:
        t = max(2.0, math.log2(n)) if n > 1 else 2.0
    rand = (rng.random if rng is not None else random.random)
    if spanner == "greedy":
        bundle, off = spanner_bundle(n, edges, t, bundles)
    else:
        raise ValueError(f"unknown spanner {spanner!r}; only 'greedy' so far "
                         "(Baswana-Sen is the scale/production option)")
    m, S = len(edges), len(bundle)
    if m <= S or not off:                       # bundle already ~ whole graph
        return list(edges), 0.0
    p = min(1.0, max(0.0, (keep_frac * m - S) / (m - S)))
    kept = list(bundle) + [(u, v, w / p) for (u, v, w) in off
                           if p > 0.0 and rand() < p]
    return kept, p


# ---------------------------------------------------------------------------
# scipy <-> edge-list bridge (SDD operators: L(graph) + diag(slack), slack >= 0)
# ---------------------------------------------------------------------------
def edges_of(A):
    """Off-diagonal edges (u, v, w) with u < v, w = -A[u, v] > 0, from a sparse
    SDD/Laplacian operator. Explicit zeros are skipped (they are not edges)."""
    T = sp.triu(sp.csr_matrix(A), k=1).tocoo()
    return [(int(i), int(j), float(-v))
            for i, j, v in zip(T.row, T.col, T.data) if v != 0.0]


def slack_of(A):
    """Per-row SDD slack: diag(A) - (sum of incident conductances). >= 0 for an
    SDD operator; the diagonal excess over the graph Laplacian."""
    A = sp.csr_matrix(A)
    d = A.diagonal()
    offdiag = A - sp.diags(d)
    off_rowsum = -np.asarray(offdiag.sum(axis=1)).ravel()   # sum of conductances
    return d - off_rowsum


def sdd_from_edges(n, edges, slack=0.0):
    """Build the SDD operator L(edges) + diag(slack) as canonical CSR float64.
    `slack` may be a scalar or a length-n vector (e.g. `slack_of(A)`)."""
    if edges:
        u = np.fromiter((e[0] for e in edges), dtype=np.int64, count=len(edges))
        v = np.fromiter((e[1] for e in edges), dtype=np.int64, count=len(edges))
        w = np.fromiter((e[2] for e in edges), dtype=np.float64, count=len(edges))
        rows = np.concatenate([u, v])
        cols = np.concatenate([v, u])
        vals = np.concatenate([-w, -w])
        W = sp.csr_matrix((vals, (rows, cols)), shape=(n, n))
        deg = np.asarray(-W.sum(axis=1)).ravel()            # sum of conductances
    else:
        W = sp.csr_matrix((n, n))
        deg = np.zeros(n)
    slack = np.full(n, float(slack)) if np.isscalar(slack) else np.asarray(slack)
    A = (W + sp.diags(deg + slack)).tocsr()
    A.sum_duplicates()
    A.sort_indices()
    return A
